api/stream.go

```diff
@@ -124,23 +118,90 @@ func (s *StreamAPI) newSubscription(
-		}
+	return s.newSubscription(
+		ctx,
+		s.blocksBroadcaster,
+		func(ctx context.Context, height uint64) (interface{}, error) {
+			block, err := s.blocks.GetByHeight(height)
+			if err != nil {
+				if errors.Is(err, storageErrs.ErrNotFound) { // make sure to wrap in not found error as the streamer expects it
+					return nil, errors.Join(flowgoStorage.ErrNotFound, err)
+				}
+				return nil, fmt.Errorf("failed to get block at height: %d: %w", height, err)
+			}
 
-		h, err := block.Hash()
-		if err != nil {
-			return nil, err
-		}
-		// todo there is a lot of missing data: https://docs.chainstack.com/reference/ethereum-native-subscribe-newheads
-		return Block{
-			Hash:         h,
-			Number:       hexutil.Uint64(block.Height),
-			ParentHash:   block.ParentBlockHash,
-			ReceiptsRoot: block.ReceiptRoot,
-			Transactions: block.TransactionHashes,
-		}, nil
-	})
+			h, err := block.Hash()
+			if err != nil {
+				return nil, errs.ErrInternal
+			}
+			// todo there is a lot of missing data: https://docs.chainstack.com/reference/ethereum-native-subscribe-newheads
+			return Block{
+				Hash:         h,
+				Number:       hexutil.Uint64(block.Height),
+				ParentHash:   block.ParentBlockHash,
+				ReceiptsRoot: block.ReceiptRoot,
+				Transactions: block.TransactionHashes}, nil
+		},
+	)
+}
+
+// NewPendingTransactions creates a subscription that is triggered each time a
+// transaction enters the transaction pool. If fullTx is true the full tx is
+// sent to the client, otherwise the hash is sent.
+func (s *StreamAPI) NewPendingTransactions(ctx context.Context, fullTx *bool) (*rpc.Subscription, error) {
+	return s.newSubscription(
+		ctx,
+		s.transactionsBroadcaster,
+		func(ctx context.Context, height uint64) (interface{}, error) {
+			block, err := s.blocks.GetByHeight(height)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get block at height: %d: %w", height, err)
+			}
+
+			// todo once a block can contain multiple transactions this needs to be refactored
+			if len(block.TransactionHashes) != 1 {
+				return nil, fmt.Errorf("block contains more than a single transaction")
+			}
+			hash := block.TransactionHashes[0]
+
+			tx, err := s.transactions.Get(hash)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get tx with hash: %s at height: %d: %w", hash, height, err)
+			}
+
+			rcp, err := s.receipts.GetByTransactionID(hash)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get receipt with hash: %s at height: %d: %w", hash, height, err)
+			}
+
+			from, err := tx.From()
+			if err != nil {
+				return nil, errs.ErrInternal
+			}
+
+			h, err := tx.Hash()
+			if err != nil {
+				return nil, err
+			}
+
+			v, r, ss := tx.RawSignatureValues()
+
+			return &RPCTransaction{
+				Hash:        h,
+				BlockHash:   &rcp.BlockHash,
+				BlockNumber: (*hexutil.Big)(rcp.BlockNumber),
+				From:        from.Hex(),
+				To:          tx.To().Hex(),
```

Comment:
I think there's a possible nil pointer dereference, because tx.To() returns a value of type: *common.Address.